### DISCUSSION TITLE
[x] Sharing computational techniques across domains

### DISCUSSION THREAD

**Discussion by manodeep** [RSEAA/RSEAA.github.io#56] [Discussion topics for #RSEAA2023]
Inspired by #51.  Is it worthwhile to share what our domains do when solving computational problems? If we categorise the computational task and then state the practices associated within individual domains, then perhaps we can develop ways to share techniques across domains. 

The idea is to either list a kind of "research" task (where practices/tools likely are domain-specific), or a computing paradigm (which should be domain agnostic but some domains may be more mature than others)

# Research Task

## Finding objects of interest in images
- Astronomy:
- BioInformatics:
- CS:
- HASS:
- 

## Finding short-lived (transient) events of interest in a series of images
- Astronomy:
- BioInformatics:
- CS:
- HASS:
- 

## Finding correlations/modulations between time-series data
- Astronomy:
- BioInformatics:
- CS:
- HASS:
- 

## Finding "close-by" pairs of objects given one or two sets of objects
- Astronomy:
- BioInformatics:
- CS:
- HASS:
- 

## Removing "noise" from an image
- Astronomy:
- BioInformatics:
- CS:
- HASS:
- 

## Evolving a system of objects under some physical/empirical laws
- Astronomy:
- BioInformatics:
- CS:
- HASS:
- 

-----

# Computational Paradigm

**Comment by pr4deepr** [comment]
For image-related tasks, 
I would add "Bioimage informatics" for image analysis in biomedical research. It is quite different from bioinformatics..

**Comment by safayeth** [comment]
Computational analysis of medical images can allow the discovery of disease patterns and correlations among cohorts of medical image data analysis along with Bioinformatics and Genetic Engineering

**Comment by robmoss** [comment]
## Model inference

There are many different approaches for fitting models to data. Suitable options can depend upon the nature of the model, the nature of the data, and the nature of the (known or assumed) relationship between the data and the model. For example, within Bayesian inference there families of methods that depend on whether you can write down the likelihood function exactly and, if so, how expensive it is to evaluate.

Annoyingly, a single method can be known by different names in different fields. One example is Sequential Monte Carlo (SMC), which is also known as Particle Filtering (PF). So it can be difficult to even identify that people in another area are doing something similar (or even the same!) without reading through all of the technical details.

**Reply by robmoss** [reply]
## Making code (somewhat) generic

There's an art in separating the "what to do" from the "how to do it", so that anyone (including yourself) can quickly adapt and apply your code to a similar or related problem. As I understand it, bioinformatics research generally has a very clear pipeline. While this may not exactly apply to other fields (e.g., perhaps there are steps where bespoke tools or code is needed) the overall approach is extremely valuable. I also find it's a helpful thing to do when trying to achieve some of the other tasks, such as making your code faster/portable and running it on HPC platforms.

**Reply by supernord** [reply]
I've always found the concept of abstracting away the code itself from the execution quite interesting
i.e. workflow management systems

**Reply by SaranjeetKaur** [reply]
@robmoss - I can so much relate to "a single method can be known by different names in different fields". It is surprising to discover that similar concepts are being used in different fields, but that mostly only happens when you get into the technical details!

**Reply by supernord** [reply]
and how do we share the ability to use these systems across domains?
e.g. Nextflow Tower, workflow execution services...

**Reply by robmoss** [reply]
I've spent a lot of time in the past few years doing this for my own inference package. One major motivation was that we were doing a huge amount of COVID-19 work for government. The knowledge, assumptions, and data systems were rapidly evolving and I wanted to do all I could to ensure I didn't break anything and unknowingly generate invalid results.

**Reply by manodeep** [reply]
Some folks used my Astronomy-specific correlation function code to study earthquakes! I don't know how they figured out that Corrfunc would do the job

**Comment by supernord** [comment]
## some suggestions for bioinformatics

- **Finding objects of interest in images:** microscopy, medical imaging, molecular imaging with mass spectrometry
- **Finding correlations/modulations between time-series data:** quantitative omics (proteomics, for example) with samples across time
- **Finding "close-by" pairs of objects given one or two sets of objects:** object as data from biological measurement (e.g. assay, MS) and using distance as a metric for change
- **Removing "noise" from an image:** de-noising of molecular image data

**Reply by supernord** [reply]
computationally intensive methods in bioinformatics include those which need to reconstruct data (e.g. genome assembly, protein identification in mass spec), which can consume a lot of memory and be quite I/O intensive

**Reply by supernord** [reply]
and individual tasks (i.e. filter data, reconstruct data pieces, polish reconstructed data object, validate reconstruction, perform QC) will be embedded as part of a given piece of software, which can then be a modular part of computational workflow

**Comment by SaranjeetKaur** [comment]
I come from Statistics background. When doing a project on time series forecasting, I had come across the an algorithm called Hamiltonian Monte Carlo - which I realised comes from Physics. 
It was great to see the connections between Statistics and Physics through this!

**Reply by SaranjeetKaur** [reply]
I would also like to share about Nested Sampling which is used in astronomical data analysis (for multi-modal data).

**Comment by manodeep** [comment]
# Domain: Astronomy

- Finding objects of interest in images: When you detect 5 photons in a CCD pixel - how do we know whether that's real or not? This is particularly relevant for objects at really early times. 

- Finding short-lived (transient) events of interest in a series of images: Some astrophysical events are very short - gamma ray bursts, supernovae etc. Images are "blink"ed to find out whether something real occurred; multi-wavelengths coverage helps.


- Finding correlations/modulations between time-series data: Used a lot in exo-planet-searches by looking for radial-velocity patterns - which are changes in the amount of starlight received because of the motion of the star + planet system. Typically uses Fourier techniques (periodogram analysis), but also time-series autocorrelations to find periods. 

- Finding "close-by" pairs of objects given one or two sets of objects: @manodeep This is my speciality, particularly when you are only interested in close spatial ranges compared to the actual extent occupied by particles. Uses some spatial partitioning methods (tree/grids) to bin-up the objects, and then only search in a much smaller volume for actual pairs. Here's a tool I created: https://github.com/manodeep/Corrfunc

- Removing "noise" from an image

- Evolving a system of objects under some physical/empirical laws: The gravitational N-body

**Comment by smanosmxm** [comment]
**some suggestions on computational paradigms**

- these are all very complex topics that require knowledge of the infrastructure, deployment / queueing systems, the programming language(s) used, and the actual scientific method being supported or implemented
- i see that any one method/package/approach that needs to be supported needs to resolve all these challenges simultaneously - which is hard for an individual
- so you need to start thinking support models and expertise and its relationship to the user.

**Reply by supernord** [reply]
also making the right aspects of computational research cross-cutting
if a field is heavily using a particular workflow management system, wouldn't it be great if this was visible

**Reply by robmoss** [reply]
I agree that this is really important and really hard. I've seen a general tendency for people to implement their own (limited, incomplete, possibly incorrect) solutions — either because they're not aware of existing solutions, or because it seems like less effort to implement a custom solution rather than learn how to use something new — and then use them for everything, regardless of how (in)appropriate it may be.

**Reply by robmoss** [reply]
@supernord made some great points in the discussion about sharing awareness of methods and approaches in different disciplines, so that we can at least be aware of other disciplines, communities, etc, who might have useful knowledge.

**Reply by robmoss** [reply]
Someone else (apologies, I forget who it was!) mentioned the value in having high-level, general examples or documentation for tools and packages that are widely-used within a field, but are likely of value beyond that field.

**Reply by pr4deepr** [reply]
I like this.
How would this look like? Are there any examples?

**Comment by SaranjeetKaur** [comment]
one suggestion from Johan Gustafsson - involve application specialists

**Reply by supernord** [reply]
particularly those that work at compute infrastructures, or who are familiar with many infrastructure types (HPC, cloud)

**Comment by SaranjeetKaur** [comment]
Propose this topic to SSI, ReSA, RDA to develop a resource or suggest if any such resource already exists!
